// @switchboard-xyz/on-demand/dist/accounts/permission.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Permission = exports.SwitchboardPermission = void 0;
var SwitchboardPermission;
(function (SwitchboardPermission) {
    SwitchboardPermission[SwitchboardPermission["PermitOracleHeartbeat"] = 1] = "PermitOracleHeartbeat";
    SwitchboardPermission[SwitchboardPermission["PermitOracleQueueUsage"] = 2] = "PermitOracleQueueUsage";
})(SwitchboardPermission || (exports.SwitchboardPermission = SwitchboardPermission = {}));
class Permission {
    static async setIx(program, params) {
        const payer = program.provider.wallet.payer;
        const ix = await program.instruction.permissionSet({
            enable: params.enable ?? false,
            permission: params.permission,
        }, {
            accounts: {
                granter: params.granter,
                authority: params.authority,
            },
            remainingAccounts: [
                { pubkey: params.grantee, isSigner: false, isWritable: true },
            ],
            signers: [payer],
        });
        return ix;
    }
    constructor() { }
}
exports.Permission = Permission;
